packages/kotti-ui/source/kotti-field-file-upload/KtFieldFileUpload.vue

```diff
@@ -0,0 +1,246 @@
+
+export default defineComponent<KottiFieldFileUpload.PropsInternal>({
+	name: 'KtFieldFileUpload',
+	components: {
+		DropArea,
+		FileItem,
+		KtField,
+		PreUploadedFileItem,
+		TakePhoto,
+	},
+	props: makeProps(KottiFieldFileUpload.propsSchema),
+	setup(props, { emit }) {
+		const field = useField<KottiFieldFileUpload.ValueInternal>({
+			emit,
+			/**
+			 * We consider this field empty if:
+			 * a. No files are selected, or
+			 * b. All selected files are invalid
+			 */
+			isEmpty: (value) =>
+				value.length === 0 ||
+				value.every(
+					(fileItem) =>
+						validateFile({
+							extensions: props.extensions,
+							fileName: 'file' in fileItem ? fileItem.file.name : fileItem.name,
+							fileSize: 'file' in fileItem ? fileItem.file.size : fileItem.size,
+							maxFileSize: props.maxFileSize,
+						}) !== KottiFieldFileUpload.Validation.SUCCESS,
+				),
+			props,
+			supports: KOTTI_FIELD_FILE_UPLOAD_SUPPORTS,
+		})
+
+		const newUploadedFiles = new Set<KottiFieldFileUpload.FileInfo['id']>()
+		const preUploadedFiles = new Set<KottiFieldFileUpload.FileInfo['id']>()
+
+		const filesList = computed<KottiFieldFileUpload.FileInfo[]>(() =>
+			field.currentValue
+				.filter((fileItem) => fileItem.id && !preUploadedFiles.has(fileItem.id))
+				.map((fileItem) => {
+					const fileInfo = buildFileInfo({
+						extensions: props.extensions,
+						fileItem,
+						isRemoteUpload: Boolean(props.remoteUpload?.actions),
+						isUploaded: newUploadedFiles.has(fileItem.id),
+						maxFileSize: props.maxFileSize,
+						progress: props.remoteUpload?.payload[fileItem.id]?.progress ?? 0,
+					})
+
+					if (fileItem.status !== fileInfo.status)
+						setStatus({ id: fileItem.id, status: fileInfo.status })
+
+					return fileInfo
+				}),
+		)
+		const showDropArea = computed(
+			() =>
+				props.allowMultiple ||
+				(!props.allowMultiple && field.currentValue.length === 0),
+		)
+
+		const setStatus = (payload: KottiFieldFileUpload.Events.SetStatus) => {
+			field.setValue(
+				field.currentValue.map((fileItem) =>
+					fileItem.id === payload.id
+						? { ...fileItem, status: payload.status }
+						: fileItem,
+				),
+			)
+
+			if (payload.status === KottiFieldFileUpload.Status.Common.UPLOADED)
+				newUploadedFiles.add(payload.id)
+		}
+
+		onBeforeMount(() => {
+			const parsedValue = KottiFieldFileUpload.valuesSchema.parse(
+				field.currentValue,
+			)
+
+			parsedValue.forEach((fileItem) => {
+				if (!isValidUploadedFileStatus(fileItem))
+					throw new Error(
+						`KtFieldFileUpload: invalid status for file item ${fileItem.id}: ${fileItem.status}`,
+					)
+				preUploadedFiles.add(fileItem.id)
+			})
+
+			field.setValue(parsedValue)
+		})
+
+		watch(
+			() => props.remoteUpload?.payload,
+			(payload) => {
+				if (!payload || isEmpty(payload)) return
+
+				field.currentValue.forEach((fileItem) => {
+					if (
+						!(
+							fileItem.status === KottiFieldFileUpload.Status.Remote.CANCELED ||
```

Comment:
I understand if you want to skip the setStatus as long as it's uploading

but if you skip the Error and Cancelled, how would they be set from outside the first time..

the first time being when this watcher triggers where newPayload is Canceled but oldPayload is Uploading

Reply:
IMO, unless I'm missing something

you'd only ignore the cancelled and error if they are set AGAIN

you need both the old and new value in the watcher to be able to tell though